File: erpnext_amex/amex_integration/page/amex_review/amex_review.py
# Copyright (c) 2025, Your Company and contributors
# For license information, please see license.txt

import frappe
from frappe import _
import json
from erpnext_amex.utils.classification_memory import get_classification_suggestion, learn_from_transaction
from erpnext_amex.utils.journal_entry_creator import create_journal_entry_from_transaction, create_bulk_journal_entries


@frappe.whitelist()
def get_pending_transactions(filters=None):
	"""Get list of pending transactions for review"""
	if filters is None:
		filters = {}
	elif isinstance(filters, str):
		filters = json.loads(filters)
	
	# Build filter conditions
	conditions = ["status IN ('Pending', 'Classified')"]
	
	if filters.get('batch_id'):
		conditions.append(f"batch_id = '{filters['batch_id']}'")
	
	if filters.get('card_member'):
		conditions.append(f"card_member LIKE '%{filters['card_member']}%'")
	
	if filters.get('from_date'):
		conditions.append(f"transaction_date >= '{filters['from_date']}'")
	
	if filters.get('to_date'):
		conditions.append(f"transaction_date <= '{filters['to_date']}'")
	
	if filters.get('min_amount'):
		conditions.append(f"amount >= {filters['min_amount']}")
	
	if filters.get('max_amount'):
		conditions.append(f"amount <= {filters['max_amount']}")
	
	where_clause = " AND ".join(conditions) if conditions else "1=1"
	
	transactions = frappe.db.sql(f"""
		SELECT 
			name, transaction_date, description, card_member, 
			amount, status, vendor, expense_account, cost_center,
			reference, amex_category, is_duplicate, is_amex_payment,
			ml_confidence_score, ml_predicted_vendor
		FROM `tabAMEX Transaction`
		WHERE {where_clause}
		ORDER BY transaction_date DESC, name DESC
		LIMIT 500
	""", as_dict=True)
	
	# Get suggestions for each transaction
	for trans in transactions:
		suggestion = get_classification_suggestion(trans.description, trans.amount)
		if suggestion:
			trans['suggestion'] = suggestion
	
	return transactions


@frappe.whitelist()
def get_transaction_details(transaction_name):
	"""Get full details of a transaction"""
	transaction = frappe.get_doc('AMEX Transaction', transaction_name)
	
	# Get suggestion
	suggestion = get_classification_suggestion(transaction.description, transaction.amount)
	
	return {
		'transaction': transaction.as_dict(),
		'suggestion': suggestion
	}


@frappe.whitelist()
def classify_transaction(transaction_name, vendor=None, expense_account=None, cost_center=None, notes=None, cost_center_splits=None):
	"""
	Classify a transaction
	
	Args:
		transaction_name: Name of the transaction
		vendor: Supplier name (optional)
		expense_account: Account name
		cost_center: Cost Center name (for single allocation)
		notes: Classification notes
		cost_center_splits: List of splits for multi-center allocation
	"""
	transaction = frappe.get_doc('AMEX Transaction', transaction_name)
	
	# Update classification
	if vendor:
		transaction.vendor = vendor
	if expense_account:
		transaction.expense_account = expense_account
	if cost_center:
		transaction.cost_center = cost_center
	if notes:
		transaction.classification_notes = notes
	
	# Handle splits
	if cost_center_splits:
		transaction.cost_center_splits = []
		for split in cost_center_splits:
			transaction.append('cost_center_splits', split)
	
	transaction.classified_by = frappe.session.user
	transaction.classification_date = frappe.utils.now()
	transaction.status = 'Classified'
	transaction.save(ignore_permissions=True)
	frappe.db.commit()
	
	# Learn from this classification
	learn_from_transaction(transaction)
	
	return {'status': 'success', 'transaction': transaction.as_dict()}


@frappe.whitelist()
def approve_transaction(transaction_name):
	"""Approve a classified transaction"""
	transaction = frappe.get_doc('AMEX Transaction', transaction_name)
	transaction.approve()
	frappe.db.commit()
	
	return {'status': 'success', 'transaction': transaction.as_dict()}


@frappe.whitelist()
def post_transaction(transaction_name):
	"""Post transaction to journal entry"""
	transaction = frappe.get_doc('AMEX Transaction', transaction_name)
	je = transaction.post_to_journal_entry()
	frappe.db.commit()
	
	return {
		'status': 'success',
		'journal_entry': je.name,
		'transaction': transaction.as_dict()
	}


@frappe.whitelist()
def bulk_approve_and_post(transaction_names):
	"""Approve and post multiple transactions"""
	import json
	
	if isinstance(transaction_names, str):
		transaction_names = json.loads(transaction_names)
	
	results = {
		'approved': [],
		'posted': [],
		'errors': []
	}
	
	for trans_name in transaction_names:
		try:
			transaction = frappe.get_doc('AMEX Transaction', trans_name)
			
			# Approve if not already
			if transaction.status == 'Classified':
				transaction.approve()
				results['approved'].append(trans_name)
			
			# Post if approved
			if transaction.status == 'Approved':
				je = transaction.post_to_journal_entry()
				results['posted'].append({'transaction': trans_name, 'journal_entry': je.name})
		
		except Exception as e:
			results['errors'].append({'transaction': trans_name, 'error': str(e)})
	
	frappe.db.commit()
	return results


@frappe.whitelist()
def create_vendor_quick(vendor_name, supplier_group=None, country=None):
	"""Quick create a vendor/supplier"""
	supplier = frappe.get_doc({
		'doctype': 'Supplier',
		'supplier_name': vendor_name,
		'supplier_group': supplier_group or 'All Supplier Groups',
		'country': country or frappe.defaults.get_user_default('country') or 'United States'
	})
	supplier.insert(ignore_permissions=True)
	frappe.db.commit()
	
	return {'status': 'success', 'supplier': supplier.name}


@frappe.whitelist()
def get_filter_options():
	"""Get options for filters"""
	batches = frappe.get_all('AMEX Import Batch', fields=['name', 'import_date'], order_by='import_date desc', limit=50)
	
	card_members = frappe.db.sql("""
		SELECT DISTINCT card_member 
		FROM `tabAMEX Transaction`
		ORDER BY card_member
	""", as_dict=True)
	
	return {
		'batches': batches,
		'card_members': [cm['card_member'] for cm in card_members]
	}


@frappe.whitelist()
def mark_as_duplicate(transaction_name, original_reference=None):
	"""Mark transaction as duplicate"""
	transaction = frappe.get_doc('AMEX Transaction', transaction_name)
	transaction.is_duplicate = 1
	transaction.duplicate_reference = original_reference
	transaction.status = 'Duplicate'
	transaction.save(ignore_permissions=True)
	frappe.db.commit()
	
	return {'status': 'success'}


@frappe.whitelist()
def get_account_list(account_type=None):
	"""Get list of accounts for dropdown"""
	filters = {}
	if account_type:
		filters['account_type'] = account_type
	
	accounts = frappe.get_all('Account', 
		filters=filters,
		fields=['name', 'account_name', 'account_type'],
		order_by='name'
	)
	
	return accounts


@frappe.whitelist()
def get_cost_center_list():
	"""Get list of cost centers for dropdown"""
	cost_centers = frappe.get_all('Cost Center',
		fields=['name', 'cost_center_name', 'parent_cost_center'],
		order_by='name'
	)
	
	return cost_centers


@frappe.whitelist()
def get_supplier_list():
	"""Get list of suppliers for dropdown"""
	suppliers = frappe.get_all('Supplier',
		fields=['name', 'supplier_name'],
		order_by='supplier_name',
		limit=1000
	)
	
	return suppliers

File: erpnext_amex/amex_integration/page/amex_review/amex_review.js
frappe.pages['amex_review'].on_page_load = function(wrapper) {
	var page = frappe.ui.make_app_page({
		parent: wrapper,
		title: 'AMEX Transaction Review',
		single_column: true
	});

	// Add page buttons
	page.add_button('Bulk Approve & Post', () => {
		if (window.amex_review_instance) {
			window.amex_review_instance.bulk_approve_and_post();
		}
	}, {icon: 'fa fa-check'});

	page.add_button('Refresh', () => {
		if (window.amex_review_instance) {
			window.amex_review_instance.load_transactions();
		}
	}, {icon: 'fa fa-refresh'});

	// Initialize the page
	window.amex_review_instance = new AMEXReviewPage(page);
}

class AMEXReviewPage {
	constructor(page) {
		this.page = page;
		this.transactions = [];
		this.selected_transaction = null;
		this.selected_transactions = new Set();
		this.keyword_debounce_timer = null;
		this.sort_field = 'transaction_date';
		this.sort_order = 'desc';
		
		// Load the HTML
		$(frappe.render_template("amex_review", {})).appendTo(this.page.body);
		
		this.setup_events();
		this.setup_autocomplete_fields();
		this.load_filter_options();
		this.load_transactions();
	}

	setup_autocomplete_fields() {
		const me = this;
		
		// Single transaction vendor field
		this.vendor_field = frappe.ui.form.make_control({
			parent: $('#vendor-field'),
			df: {
				fieldtype: 'Link',
				options: 'Supplier',
				label: 'Vendor/Supplier',
				placeholder: 'Type to search...'
			},
			render_input: true
		});
		
		// Single transaction expense account field
		this.expense_account_field = frappe.ui.form.make_control({
			parent: $('#expense-account-field'),
			df: {
				fieldtype: 'Link',
				options: 'Account',
				label: 'Expense Account',
				placeholder: 'Type to search...',
				get_query: () => {
					return {
						filters: {
							'account_type': 'Expense',
							'disabled': 0
						}
					};
				}
			},
			render_input: true
		});
		
		// Single transaction cost center field
		this.cost_center_field = frappe.ui.form.make_control({
			parent: $('#cost-center-field'),
			df: {
				fieldtype: 'Link',
				options: 'Cost Center',
				label: 'Cost Center',
				placeholder: 'Type to search...',
				get_query: () => {
					return {
						filters: {
							'disabled': 0
						}
					};
				}
			},
			render_input: true
		});
		
		// Bulk vendor field
		this.bulk_vendor_field = frappe.ui.form.make_control({
			parent: $('#bulk-vendor-field'),
			df: {
				fieldtype: 'Link',
				options: 'Supplier',
				label: 'Vendor/Supplier',
				placeholder: 'Type to search...'
			},
			render_input: true
		});
		
		// Bulk expense account field
		this.bulk_expense_account_field = frappe.ui.form.make_control({
			parent: $('#bulk-expense-account-field'),
			df: {
				fieldtype: 'Link',
				options: 'Account',
				label: 'Expense Account',
				placeholder: 'Type to search...',
				get_query: () => {
					return {
						filters: {
							'account_type': 'Expense',
							'disabled': 0
						}
					};
				}
			},
			render_input: true
		});
		
		// Bulk cost center field
		this.bulk_cost_center_field = frappe.ui.form.make_control({
			parent: $('#bulk-cost-center-field'),
			df: {
				fieldtype: 'Link',
				options: 'Cost Center',
				label: 'Cost Center',
				placeholder: 'Type to search...',
				get_query: () => {
					return {
						filters: {
							'disabled': 0
						}
					};
				}
			},
			render_input: true
		});
	}

	setup_events() {
		const me = this;

		// Apply filters
		$('#apply-filters-btn').click(() => me.load_transactions());
		
		// Keyword filter with debounce
		$('#filter-keyword').on('input', function() {
			clearTimeout(me.keyword_debounce_timer);
			me.keyword_debounce_timer = setTimeout(() => {
				me.load_transactions();
			}, 500);
		});

		// Select all checkbox
		$('#select-all-transactions').change(function() {
			const checked = $(this).is(':checked');
			$('.transaction-checkbox').prop('checked', checked);
			me.update_selected_transactions();
		});

		// Transaction row click
		$(document).on('click', '.transaction-row', function(e) {
			if ($(e.target).is('input[type="checkbox"]')) return;
			
			const name = $(this).data('name');
			me.load_transaction_details(name);
			$('.transaction-row').removeClass('table-active');
			$(this).addClass('table-active');
		});

		// Transaction checkbox
		$(document).on('change', '.transaction-checkbox', function() {
			me.update_selected_transactions();
		});

		// Sortable column headers
		$(document).on('click', '.sortable-header', function() {
			const field = $(this).data('field');
			me.sort_transactions(field);
		});

		// Cost center allocation type
		$('#single-cc-btn').click(() => me.toggle_cost_center_type('single'));
		$('#split-cc-btn').click(() => me.toggle_cost_center_type('split'));

		// Add split button
		$('#add-split-btn').click(() => me.add_split_row());

		// Remove split button
		$(document).on('click', '.remove-split-btn', function() {
			$(this).closest('tr').remove();
			me.calculate_split_totals();
		});

		// Split amount/percentage change
		$(document).on('input', '.split-amount, .split-percentage', function() {
			me.calculate_split_totals();
		});

		// Classify button
		$('#classify-btn').click(() => me.classify_transaction());

		// Approve button
		$('#approve-btn').click(() => me.approve_transaction());

		// Post button
		$('#post-btn').click(() => me.post_transaction());

		// Mark duplicate button
		$('#mark-duplicate-btn').click(() => me.mark_as_duplicate());
		
		// Bulk classify button
		$('#bulk-classify-btn').click(() => me.bulk_classify_transactions());

		// Create vendor button
		$('#create-vendor-btn').click(() => $('#vendor-modal').modal('show'));

		// Save vendor button
		$('#save-vendor-btn').click(() => me.create_vendor());
	}

	load_filter_options() {
		frappe.call({
			method: 'erpnext_amex.amex_integration.page.amex_review.amex_review.get_filter_options',
			callback: (r) => {
				if (r.message) {
					// Populate batch filter
					r.message.batches.forEach(batch => {
						$('#filter-batch').append(`<option value="${batch.name}">${batch.name} (${batch.import_date})</option>`);
					});

					// Populate card member filter
					r.message.card_members.forEach(member => {
						$('#filter-card-member').append(`<option value="${member}">${member}</option>`);
					});
				}
			}
		});
	}

	load_transactions() {
		const me = this;
		const filters = {
			batch_id: $('#filter-batch').val(),
			card_member: $('#filter-card-member').val(),
			from_date: $('#filter-from-date').val(),
			to_date: $('#filter-to-date').val(),
			keyword: $('#filter-keyword').val()
		};

		$('#loading-transactions').show();
		$('#no-transactions').hide();
		$('#transaction-list').empty();

		frappe.call({
			method: 'erpnext_amex.amex_integration.page.amex_review.amex_review.get_pending_transactions',
			args: { filters: JSON.stringify(filters) },
			callback: (r) => {
				$('#loading-transactions').hide();
				
				if (r.message && r.message.length > 0) {
					me.transactions = r.message;
					me.render_transactions();
				} else {
					$('#no-transactions').show();
				}
			}
		});
	}

	sort_transactions(field) {
		// Toggle sort order if clicking same field
		if (this.sort_field === field) {
			this.sort_order = this.sort_order === 'asc' ? 'desc' : 'asc';
		} else {
			this.sort_field = field;
			this.sort_order = 'asc';
		}

		// Sort transactions
		this.transactions.sort((a, b) => {
			let aVal = a[field];
			let bVal = b[field];

			// Handle nulls
			if (aVal === null || aVal === undefined) aVal = '';
			if (bVal === null || bVal === undefined) bVal = '';

			// Compare based on type
			if (field === 'amount') {
				aVal = Number(aVal);
				bVal = Number(bVal);
			} else if (field === 'transaction_date') {
				aVal = new Date(aVal);
				bVal = new Date(bVal);
			} else {
				aVal = String(aVal).toLowerCase();
				bVal = String(bVal).toLowerCase();
			}

			if (this.sort_order === 'asc') {
				return aVal > bVal ? 1 : -1;
			} else {
				return aVal < bVal ? 1 : -1;
			}
		});

		this.render_transactions();
		this.update_sort_indicators();
	}

	update_sort_indicators() {
		// Remove all sort indicators
		$('.sortable-header .sort-indicator').remove();

		// Add indicator to current sorted column
		const icon = this.sort_order === 'asc' ? 'fa-sort-asc' : 'fa-sort-desc';
		$(`.sortable-header[data-field="${this.sort_field}"]`)
			.append(`<i class="fa ${icon} sort-indicator" style="margin-left: 5px;"></i>`);
	}

	render_transactions() {
		const tbody = $('#transaction-list');
		tbody.empty();

		// Calculate total pending amount
		let total_pending = 0;
		this.transactions.forEach(trans => {
			if (trans.status === 'Pending' || trans.status === 'Classified') {
				total_pending += Number(trans.amount) || 0;
			}
		});

		// Update total display
		$('#total-pending-amount').text(`$${total_pending.toFixed(2)}`);
		$('#total-pending-count').text(this.transactions.length);

		this.transactions.forEach(trans => {
			const statusClass = {
				'Pending': 'badge-warning',
				'Classified': 'badge-info',
				'Approved': 'badge-success',
				'Posted': 'badge-secondary'
			}[trans.status] || 'badge-secondary';

			const row = `
				<tr class="transaction-row" data-name="${trans.name}">
					<td>
						<input type="checkbox" class="transaction-checkbox" value="${trans.name}">
					</td>
					<td>${frappe.datetime.str_to_user(trans.transaction_date)}</td>
					<td>${trans.description || ''}</td>
					<td>${trans.card_member || ''}</td>
					<td>$${Number(trans.amount).toFixed(2)}</td>
					<td><span class="badge ${statusClass}">${trans.status}</span></td>
				</tr>
			`;
			tbody.append(row);
		});

		this.update_sort_indicators();
	}

	update_selected_transactions() {
		const me = this;
		this.selected_transactions.clear();
		
		$('.transaction-checkbox:checked').each(function() {
			me.selected_transactions.add($(this).val());
		});

		const count = this.selected_transactions.size;
		$('#selected-count').text(count);

		if (count > 1) {
			$('#bulk-panel').show();
			$('#classification-panel').hide();
		} else if (count === 1) {
			$('#bulk-panel').hide();
			$('#classification-panel').show();
		} else {
			$('#bulk-panel').hide();
			$('#classification-panel').hide();
		}
	}

	bulk_classify_transactions() {
		const me = this;
		const transaction_names = Array.from(this.selected_transactions);
		
		if (transaction_names.length === 0) {
			frappe.msgprint('Please select transactions to classify');
			return;
		}

		const vendor = me.bulk_vendor_field.get_value();
		const expense_account = me.bulk_expense_account_field.get_value();
		const cost_center = me.bulk_cost_center_field.get_value();
		const notes = $('#bulk-classification-notes').val();

		if (!expense_account) {
			frappe.msgprint('Please select an Expense Account');
			return;
		}

		if (!cost_center) {
			frappe.msgprint('Please select a Cost Center');
			return;
		}

		frappe.call({
			method: 'erpnext_amex.amex_integration.page.amex_review.amex_review.bulk_classify_transactions',
			args: {
				transaction_names: JSON.stringify(transaction_names),
				vendor: vendor,
				expense_account: expense_account,
				cost_center: cost_center,
				notes: notes
			},
			callback: (r) => {
				if (r.message) {
					const result = r.message;
					frappe.msgprint(`
						<strong>Bulk Classification Complete</strong><br>
						Success: ${result.success_count}<br>
						Errors: ${result.error_count}<br>
						Total: ${result.total}
					`);
					
					// Clear selections and reload
					me.selected_transactions.clear();
					$('.transaction-checkbox').prop('checked', false);
					$('#select-all-transactions').prop('checked', false);
					me.update_selected_transactions();
					me.load_transactions();
				}
			}
		});
	}

	load_transaction_details(transaction_name) {
		const me = this;
		me.selected_transaction = transaction_name;

		frappe.call({
			method: 'erpnext_amex.amex_integration.page.amex_review.amex_review.get_transaction_details',
			args: { transaction_name: transaction_name },
			callback: (r) => {
				if (r.message) {
					me.render_transaction_details(r.message.transaction);
					$('#classification-panel').show();
				}
			}
		});
	}

	render_transaction_details(trans) {
		const detailsHtml = `
			<div class="detail-row">
				<span class="detail-label">Reference:</span>
				<span class="detail-value">${trans.reference || 'N/A'}</span>
			</div>
			<div class="detail-row">
				<span class="detail-label">Date:</span>
				<span class="detail-value">${frappe.datetime.str_to_user(trans.transaction_date)}</span>
			</div>
			<div class="detail-row">
				<span class="detail-label">Card Member:</span>
				<span class="detail-value">${trans.card_member || 'N/A'}</span>
			</div>
			<div class="detail-row">
				<span class="detail-label">Amount:</span>
				<span class="detail-value amount-large">$${Number(trans.amount).toFixed(2)}</span>
			</div>
			<div class="detail-row">
				<span class="detail-label">Category:</span>
				<span class="detail-value">${trans.amex_category || 'N/A'}</span>
			</div>
		`;
		$('#transaction-details').html(detailsHtml);

		// Set current values if classified
		if (trans.vendor) {
			this.vendor_field.set_value(trans.vendor);
		}
		if (trans.expense_account) {
			this.expense_account_field.set_value(trans.expense_account);
		}
		if (trans.cost_center) {
			this.cost_center_field.set_value(trans.cost_center);
		}
		if (trans.classification_notes) {
			$('#classification-notes').val(trans.classification_notes);
		}
	}

	toggle_cost_center_type(type) {
		if (type === 'single') {
			$('#single-cc-btn').addClass('active');
			$('#split-cc-btn').removeClass('active');
			$('#single-cost-center-div').show();
			$('#split-cost-centers-div').hide();
		} else {
			$('#single-cc-btn').removeClass('active');
			$('#split-cc-btn').addClass('active');
			$('#single-cost-center-div').hide();
			$('#split-cost-centers-div').show();
		}
	}

	classify_transaction() {
		const me = this;
		const vendor = me.vendor_field.get_value();
		const expense_account = me.expense_account_field.get_value();
		const cost_center = me.cost_center_field.get_value();
		const notes = $('#classification-notes').val();

		if (!expense_account) {
			frappe.msgprint('Please select an Expense Account');
			return;
		}

		frappe.call({
			method: 'erpnext_amex.amex_integration.page.amex_review.amex_review.classify_transaction',
			args: {
				transaction_name: me.selected_transaction,
				vendor: vendor,
				expense_account: expense_account,
				cost_center: cost_center,
				notes: notes
			},
			callback: (r) => {
				if (r.message) {
					frappe.show_alert({message: 'Transaction classified', indicator: 'green'});
					me.load_transactions();
				}
			}
		});
	}

	approve_transaction() {
		// Stub for future implementation
		frappe.msgprint('Approve functionality coming soon');
	}

	post_transaction() {
		const me = this;
		frappe.call({
			method: 'erpnext_amex.amex_integration.page.amex_review.amex_review.post_transaction',
			args: { transaction_name: me.selected_transaction },
			callback: (r) => {
				if (r.message) {
					frappe.show_alert({message: 'Posted to Journal Entry', indicator: 'green'});
					me.load_transactions();
					$('#classification-panel').hide();
				}
			}
		});
	}

	mark_as_duplicate() {
		const me = this;
		frappe.call({
			method: 'erpnext_amex.amex_integration.page.amex_review.amex_review.mark_as_duplicate',
			args: { transaction_name: me.selected_transaction },
			callback: (r) => {
				if (r.message) {
					frappe.show_alert({message: 'Marked as duplicate', indicator: 'orange'});
					me.load_transactions();
					$('#classification-panel').hide();
				}
			}
		});
	}

	bulk_approve_and_post() {
		const me = this;
		const selected = Array.from(this.selected_transactions);
		
		if (selected.length === 0) {
			frappe.msgprint('Please select transactions to post');
			return;
		}

		frappe.confirm(
			`Post ${selected.length} transaction(s) to Journal Entries?`,
			() => {
				frappe.call({
					method: 'erpnext_amex.amex_integration.page.amex_review.amex_review.bulk_approve_and_post',
					args: { transaction_names: JSON.stringify(selected) },
					callback: (r) => {
						if (r.message) {
							frappe.msgprint(`Posted ${r.message.success_count} transactions`);
							me.selected_transactions.clear();
							me.load_transactions();
						}
					}
				});
			}
		);
	}

	create_vendor() {
		const vendor_name = $('#new-vendor-name').val();
		const supplier_group = $('#new-vendor-group').val();
		const country = $('#new-vendor-country').val();

		if (!vendor_name) {
			frappe.msgprint('Please enter vendor name');
			return;
		}

		frappe.call({
			method: 'erpnext_amex.amex_integration.page.amex_review.amex_review.create_vendor_quick',
			args: {
				vendor_name: vendor_name,
				supplier_group: supplier_group,
				country: country
			},
			callback: (r) => {
				if (r.message) {
					frappe.show_alert({message: 'Vendor created', indicator: 'green'});
					$('#vendor-modal').modal('hide');
					this.vendor_field.set_value(r.message.name);
				}
			}
		});
	}
}

